misc/userspace-task-models.md

```diff
@@ -0,0 +1,240 @@
+# Userspace Task Models
+
+## Introduction
+
+A *task* is a fundamental concept in scheduling, representing some amount of
+work to be performed. But how we think about and model tasks can vary.  This
+document explores various userspace[<sup>1</sup>](#notes) task models, how they map to
+the browser's event loop, and how various scheduling APIs fit in.
+
+## Userspace Tasks
+
+A *userspace task* is JavaScript code that performs some amount of work, e.g.
+fetch and display search results in response to a button click, render a
+framework's virtual DOM, or hydrate all or part of a page. It's up to the web
+developer as to how an application is broken into tasks.
+
+Userspace tasks:
+
+ * **Have an entry point**. This is commonly a callback, e.g. the one passed to
+   `scheduler.postTask()` or `setTimeout()`, but can also be an event listener,
+   a microtask[<sup>2</sup>](#notes), or [something else](#multiple-entry-points).
+
+ * **Can be [synchronous or asynchronous](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Introducing)**.
+   We define an *asynchronous task* as a task with asynchronous control flow, i.e.
+   it has an  *asynchronous hop* or *continuation*, such as a promise
+   continuation or async callback. Synchronous tasks do not have asynchronous
+   hops.
+
+ * **Can be yieldy**. Async tasks are *yieldy* if they yield to the event loop
+   during their execution, either by interacting with a yieldy async API, e.g.
+   `fetch()`, or by purposefully yielding, e.g. scheduling continuations with
+   `setTimeout()` or `postTask()`. Note that async tasks are not necessarily
+   yieldy.
```

Comment:
This is slightly surprising. The only example I can think of is a userspace task which is broken up by posting a microtask to continue its work. Are there other types of asynchronous, non-yieldy userspace tasks?

Reply:
Yeah OK, that makes sense. I'd suggest "Note that async tasks are not necessarily yieldy; e.g. `await`ing an already-fulfilled promise makes your task async, but will not yield".

Reply:
Done. There's also a section later on non-yieldy async tasks with an example.